nailgun/entity_mixins.py

```diff
@@ -931,3 +937,325 @@ def update(self, fields=None):
+        * No other method should alter ``fields`` or ``query``.
+
+        :param results: A list of dicts, where each dict is a set of attributes
+            for one entity. The contents of these dicts are as is returned from
+            the server.
+        :returns: A list of dicts, where each dict is a set of attributes for
+            one entity. The contents of these dicts have been normalized and
+            can be used to instantiate entities.
+
+        """
+        fields = self.get_fields()
+        normalized = []
+        for result in results:
+            # For each field that we know about, copy the corresponding field
+            # from the server's search result. If any extra attributes are
+            # copied over, Entity.__init__ will raise a NoSuchFieldError.
+            # Examples of problematic results from server:
+            #
+            # * organization_id (denormalized OneToOne. see above)
+            # * organizations, organization_ids (denormalized OneToMany. above)
+            # * updated_at, created_at (these may be handled in the future)
+            # * sp_subnet (Host.sp_subnet is an undocumented field)
+            #
+            attrs = {}
+            for field_name, field in fields.items():
+                if isinstance(field, OneToOneField):
+                    try:
+                        attrs[field_name] = _get_entity_id(field_name, result)
+                    except MissingValueError:
+                        pass
+                elif isinstance(field, OneToManyField):
+                    try:
+                        attrs[field_name] = _get_entity_ids(field_name, result)
+                    except MissingValueError:
+                        pass
+                else:
+                    try:
+                        attrs[field_name] = result[field_name]
+                    except KeyError:
+                        pass
+            normalized.append(attrs)
+        return normalized
+
+    def search(self, fields=None, query=None, filters=None):
+        """Search for entities.
+
+        At its simplest, this method searches for all entities of a given kind.
+        For example, to ask for all :class:`nailgun.entities.Organization`
+        entities::
+
+            Organization().search()
+
+        Values on an entity are used to generate a search query, and the
+        ``fields`` argument can be used to specify which fields should be used
+        when generating a search query::
+
+            org = Organization(name='foo', description='bar')
+            results = org.search()  # Search by name and description.
+            results = org.search(fields={'name', 'description'})  # Same.
+            results = org.search(fields={'name'})  # Search by name.
+            results = org.search(fields={'description'})  # Search by desc
+            results = org.search(fields=set())  # Search for all organizations.
+            results = org.search(fields={'title'})  # Error!
```

Comment:
@oshtaier be careful. the line ``org = Organization(name='foo', description='bar')``` has only created an Organization locally. At this point, the Org only has those two fields set. This will remain the case until the Org is assigned a title or is created at the server. See below for some examples.

```
In [5]: org = entities.Organization(name='foo', description='bar')

In [6]: org
Out[6]: nailgun.entities.Organization(nailgun.config.ServerConfig(url=u'https://rhsm-qe-1.rhq.lab.eng.bos.redhat.com/', auth=(u'admin', u'changeme')), description='bar', name='foo')

In [7]: org.get_values()
Out[7]: {'description': 'bar', 'name': 'foo'}

In [8]: org.get_fields()
Out[8]: 
{'compute_resource': <nailgun.entity_fields.OneToManyField at 0x7ff314bf79d0>,
 'config_template': <nailgun.entity_fields.OneToManyField at 0x7ff314bf7a10>,
 'description': <nailgun.entity_fields.StringField at 0x7ff314bf7a50>,
 'domain': <nailgun.entity_fields.OneToManyField at 0x7ff314bf7a90>,
 'environment': <nailgun.entity_fields.OneToManyField at 0x7ff314bf7ad0>,
 'hostgroup': <nailgun.entity_fields.OneToManyField at 0x7ff314bf7b10>,
 'id': <nailgun.entity_fields.IntegerField at 0x7ff314bf7e10>,
 'label': <nailgun.entity_fields.StringField at 0x7ff314bf7b50>,
 'media': <nailgun.entity_fields.OneToManyField at 0x7ff314bf7b90>,
 'name': <nailgun.entity_fields.StringField at 0x7ff314bf7bd0>,
 'realm': <nailgun.entity_fields.OneToManyField at 0x7ff314bf7c10>,
 'smart_proxy': <nailgun.entity_fields.OneToManyField at 0x7ff314bf7c50>,
 'subnet': <nailgun.entity_fields.OneToManyField at 0x7ff314bf7c90>,
 'title': <nailgun.entity_fields.StringField at 0x7ff314bf7cd0>,
 'user': <nailgun.entity_fields.OneToManyField at 0x7ff314bf7d10>}

In [9]: org.title
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-9-5f07d6148295> in <module>()
----> 1 org.title

AttributeError: 'Organization' object has no attribute 'title'
```